String Formatting Lab and Mail Room Application

shibinmathew opened the pull request.

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+def questionModule():
+    #  Function that acts as the main gate into the app
+    quit = False
+    while not quit:
```
`quit` is a built-in module that is being overwritten here

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+            optionTwo()
+        elif userResponse == 'quit':
+            print('Thanks for using this application!')
+            quit = True
```
`quit` var is not necessary since you're calling sys.exit right after it. `while True` should be sufficient.

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+            optionOne()
+        elif userResponse == "Create a Report":
+            optionTwo()
+        elif userResponse == 'quit':
```
general comment, per pep8 var names and function names are lower case -> `user_response`, `question_module`

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+    #  Function that acts as the main gate into the app
+    quit = False
+    while not quit:
+        print("Please choose from the following choics: Send a Thank You, Create a Report or quit")
```
typing out strings that are case sensitive is not really user friendly. indicate to the user to select from numeric options: 1 through 3

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+            .format(names['name'], '$', round(sum(names['donation']), 2),
+                len(names['donation']), '$',
+                round(sum(names['donation'])/len(names['donation']), 2)))
+    questionModule()
```
`questionModule` is a forever loop. there is no need to call it again.

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+        print("{: <24s}{: <2s}{: >10}{: >13}{: >6s}{: >11}"
+            .format(names['name'], '$', round(sum(names['donation']), 2),
+                len(names['donation']), '$',
+                round(sum(names['donation'])/len(names['donation']), 2)))
```
this report should be sorted per assignment spec

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+                print(donors)  # Debug
+                quit = True
+            else:
+                quit = True
```
consider refactoring above, there are too many if / else statements and too many places where quit is being kept tracked. remember that you can also use `break` statement to exit a loop.

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+            for names in donors:
+                print(names['name'])
+            print("Please select from the list or add a new name")
+            name = input()
```
`input` function accepts a string, this way instead of always having two lines this can be `input('Please select .. ')`

less unnecessary lines for reader to read improves readability of code.

natasha-aleksandrova commented on students/shibin_mathew/lesson3/mailroom.py:
```diff
@@ -0,0 +1,83 @@
+                    for names in donors:
+                        if names['name'] == name:
+                            names['donation'].append(float(donation))
+                    print(donors)  # Debug
```
this currently has a bug that if you never issue `list` command and try to add to existing donor it just duplicates the donor instead of appending.

shibinmathew commented:
Hello,

Thanks for the detailed feedback. I fixed all the pep-8 violations and
updated the code with your recommendations. I checked back at the
instruction and validated my report functionality. The assignment doesn't
specify the number of spaces in between the columns so I coded it based on
visual inspection.

The new pull request is shown below:
https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/409

I have also attached my mailroom.py file.


Thanks,

Shibin

On Wed, Apr 4, 2018 at 12:12 AM, Natasha <notifications@github.com> wrote:

> Merged #382
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/382>
> .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/382#event-1555566095>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOhKiWjhwQv1OAACRq31RSw6PDbvWE-uks5tlEgggaJpZM4TEYJ1>
> .
>

#!/usr/bin/env python3
import sys


def option_one():
    #  Function to Send a Thank You Email
    program_quit = False
    while not program_quit:
        name = input("Please input your full name\n")
        if name == "list":
            print("list of names")
            for names in donors:
                print(names['name'])
            name = input("Please select from the list or add a new name\n")
            if name != 'quit':
                donation = input("Please enter a donation amount\n")
                if donation != 'quit':
                    for names in donors:
                        if names['name'] == name:
                            names['donation'].append(float(donation))
                    print(donors)  # Debug
                    break  # break out of the while loop
                else:
                    break  # quit has been initialized
            else:
                program_quit = True
        elif name == 'quit':
            break
        else:
            donation = input("Please enter a donation amount\n")
            if donation == 'quit':
                break
            else:
                donation = float(donation)
                if not any(i['name'] == name for i in donors):
                    donors.append({'name': name, 'donation': [donation]})
                else:
                    for names in donors:
                        if names['name'] == name:
                            names['donation'].append(float(donation))
                print(donors)  # Debug
                break  # break out of the while loop


def option_two():
    #  Function to create a report
    print("hello")
    print("{: <20s}{: ^4s}{: <10s}{: ^4s}{: <10s}{: ^4s}{: <10s}"
          .format('Donor Name', '|', 'Total Given', '|',
                  'Num Gifts', '|', 'Average Gift'))
    print('-' * 70)
    print(' ')
    for names in donors:
        print("{: <24s}{: <2s}{: >10}{: >13}{: >6s}{: >11}"
              .format(names['name'], '$', round(sum(names['donation']), 2),
                      len(names['donation']), '$',
                      round(sum(names['donation'])/len(names['donation']), 2)))


def question_module():
    #  Function that acts as the main gate into the app
    program_quit = False
    while not program_quit:
        user_response = input("Please choose from the following choices: "
                              "1. Send a Thank You, 2. Create a Report, 3. quit\n")
        if user_response == "1" or user_response == "Send a Thank You":
            option_one()
        elif user_response == "2" or user_response == "Create a Report":
            option_two()
        elif user_response == "3" or user_response == "quit":
            print('Thanks for using this application!')
            sys.exit()
        else:
            print("Please pick from one of the choices")


if __name__ == "__main__":
    donors = [{'name': 'Shibin', 'donation': [25.25, 456.21]},
              {'name': 'Kimberly', 'donation': [125.50, 5.55, 14.5]},
              {'name': 'Andreck',
               'donation': [1250, 50, 1555]}, {'name': 'Jordy', 'donation':
              [55, 25, 4.50]}, {'name': 'Nelly', 'donation': [254.2, 500, 125]}]
    question_module()
